Add sniffio support

altendky opened the pull request:
https://twistedmatrix.com/trac/ticket/9815

Draft for:
- [x] newsfragment
- [x] tests
- [x] ~~verification in sniffio tests~~
  - sniffio trusts us to test
- [x] asyncioreactor tests
- [x] finish setup.py additions
- [x] update setup.py tests
- [x] update extras docs
- [ ] sniffio release
- [ ] update per modified sniffio

altendky committed:
preliminary pass at sniffio support

https://twistedmatrix.com/trac/ticket/9815

altendky committed:
only support sniffio if installed

altendky committed:
add newsfragments/9815.feature

njsmith commented:
So the tricky bit here will be that you want the contextvar to be set to `"twisted"` whenever you're inside the twisted coroutine runner, and not otherwise. In particular, if someone's using the aio reactor, then they can have asyncio `Task`s and twisted `@inlineCallbacks` code running in the same process. `sniffio` does have some fallback logic to return `"asyncio"` if there's an asyncio loop running and the contextvar is unset, so it might be enough to just set the cvar to `"twisted"` whenever entering the `inlineCallbacks` driver loop, and unset the cvar again when exiting.

altendky committed:
add sniffio tests

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

njsmith commented:
I'd probably also add a test for twisted-calling-into-asyncio-calling-into-twisted and asserting that at each layer you get the right result.

altendky commented:
First I need to figure out the first two layers...  somehow the asyncioreactor test file doesn't even have an async def anywhere to use as reference.

njsmith commented:
I think the magic search terms are fromFuture and asFuture

On Sun, May 10, 2020 at 12:21 PM Kyle Altendorf <notifications@github.com>
wrote:

> First I need to figure out the first two layers... somehow the
> asyncioreactor test file doesn't even have an async def anywhere to use as
> reference.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/twisted/twisted/pull/1263#issuecomment-626375958>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEU42BJY6QNGF66YQWB2LLRQ35CNANCNFSM4M26FU7Q>
> .
>


-- 
Nathaniel J. Smith -- https://vorpus.org <http://vorpus.org>

altendky committed:
asyncio in twisted in asyncio in twisted test of sniffio

moshez commented:
I'll note that it's more complicated than that -- we have pure-Twisted code that uses `async def`:
https://klein.readthedocs.io/en/latest/examples/await.html

njsmith commented:
@moshez what's more complicated than what? there's some test cases here that also involve pure-Twisted code using `async def`...

altendky commented:
It doesn't work yet so I'll agree that it's more complicated than what has been implemented.  `:]`  But yes, there should already be coverage of pure Twisted `async def`.

altendky committed:
camelCase

altendky committed:
reduce setting of current_async_library_cvar

altendky committed:
use an asyncio task factory to clear the cvar

altendky commented:
I'm not an immediate fan of setting the task factory to hook in and clear the context var but it does pass the additional tests introduced here.  This makes it at least a useful reference point for me to think from.  At this point I don't have any leads other than 1) force the user to call a Twisted provided wrapper of `asyncio.ensure_future()` or 2) some form of hooking into asyncio itself such as the loop's task factory.  There's some gap in my debugging where `task.Task()` doesn't end up running the expected `__init__()` and instead drops into `asyncio.base_events.BaseEventLoop.call_soon()` with the `context` parameter already having the sniffio var set to `'twisted'`.

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky commented:
@njsmith, did you have something in mind specifically about the three layers of twisted/asyncio/twisted vs. just separate twisted/asyncio and asyncio/twisted tests?  I have all of them at the moment (locally) and twisted calling asyncio did uncover an issue (thanks as always).  It seems like maybe the pair of two layer tests would be more 'unit-y' and concise and provide more distinct test results.

altendky committed:
more granular asyncio/twisted nesting tests

altendky committed:
text sniffio in exception handler in twisted coroutine

altendky committed:
expound upon news fragment

altendky committed:
skip asyncio tests sniffio test when sniffio is not available

altendky commented:
Looks like the test failures may simply be the existing contextvars support not working with the backport of contextvars.  Digging into that.

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
linting touch up

altendky committed:
fixup sniffio in setup.py and add tests

altendky committed:
add sniffio extra to optional docs

altendky committed:
(maybe) make the tests better

altendky committed:
correct to self.failureResultOf()

altendky committed:
cleanup for pyflakes

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
Remove question comments

altendky committed:
lint

glyph committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
Change sniffio import style to satisfy mypy

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
And in asyncioreactor as well

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

twm committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
black

altendky committed:
.git-blame-ignore-revs

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
we now from-import

altendky committed:
mypy

altendky committed:
Handle Windows asyncio event loop policy

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
Update defer.py

adiroiban commented on src/twisted/internet/defer.py:
```diff
@@ -52,6 +52,16 @@ def _copy_context():  # type: ignore[misc]
         return _NoContext
 
 
+current_async_library_cvar = None
```
Do we need current_async_library_cvar as public API ? If yes, then it needs documentation.

This looks like duplicate code from src/twisted/internet/asyncioreactor.py

can we do something like this here:

```
from twisted.internet.asyncioreactor import _current_async_library_cvar
```

adiroiban commented on src/twisted/internet/asyncioreactor.py:
```diff
@@ -24,7 +24,30 @@
+        pass
+
+
+def sniffioTaskFactory(loop, coro):
```
do we need this public API ?... if yes, it needs docstring.

adiroiban commented on src/twisted/internet/asyncioreactor.py:
```diff
@@ -24,7 +24,30 @@
+except ImportError:
+    _contextvarsSupport = False
+
+current_async_library_cvar = None
```
do we need it as public api?

Can we access private members in the context?

adiroiban commented on src/twisted/internet/test/test_asyncioreactor.py:
```diff
@@ -276,3 +295,150 @@ def test_noCycleReferencesInCallLater(self):
+
+
+class AsyncioSelectorReactorSniffioTests(ReactorBuilder, SynchronousTestCase):
+
```
this needs docstring...and same comment for the tests from this class.

adiroiban commented on src/twisted/test/test_defer.py:
```diff
@@ -3266,3 +3273,120 @@ async def testFunction():
+
+class CoroutineSniffioTests(unittest.TestCase):
+
+    skip = sniffioSkip
```
maybe  add dosctring to the class and use `@skipIf` decorator.

adiroiban reviewed (commented):
Thanks for the PR>

I just took a look at this PR and left some generic comments.

I have never used asyncio or trio .. so I can't do a full review of this.

The defer tests are ok... but I don't quite understand the AsyncioSelectorReactorSniffioTests

My main comments are about missing docstrings and exposed public API.

I think that if someone found it useful and want to use it we can have it in Twisted.

@altendky  if you can update it with trunk and add the missing doctring or hide the API you can ping me and I will give it another try.

Meanwhile I will try to read more about asyncio and trio and shinny new py3 stuff :)

graingert committed:
Merge branch 'trunk' of github.com:twisted/twisted into 9815-altendky-sniffio

altendky committed:
Merge pull request #2 from graingert/9815-altendky-sniffio

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
typo

graingert commented on src/twisted/internet/asyncioreactor.py:
```diff
@@ -50,6 +73,8 @@ class AsyncioSelectorReactor(PosixReactorBase):
         if eventloop is None:
             _eventloop: AbstractEventLoop = get_event_loop()
+            if current_async_library_cvar is not None:
+                _eventloop.set_task_factory(factory=sniffioTaskFactory)
```
Using this approach we'd have to also wrap  Loop.call_soon(), Loop.call_later(), and Loop.call_at() which also copy the context

altendky replied:
This strikes me as worth requesting a change so it can't be forgotten?

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

agronholm commented:
> > > Any plans on implementing this to gain AnyIO support?
> > 
> > 
> > I believe that ship has already sailed.
> 
> That's unfortunate. Do you have any insight as to why? Lack of interest? Not enough resources?

A lack of interest from the Twisted side. This was not a technical issue.

agronholm commented:
I could misremember, but I recall having actually dedicated a fair amount of time to solving this, and verdict was that without upstream support this was near impossible to implement correctly.

altendky committed:
Merge branch 'trunk' into 9815-altendky-sniffio

altendky committed:
tidy

altendky committed:
extra

altendky committed:
tidy

altendky committed:
yeah...

altendky committed:
Convert WorkerProtocol.run from async def to sync def returning Deferred

This partially reverts the change at
https://github.com/twisted/twisted/commit/a5127a50b75ca0ad4703abe91c6ad30cf85878e1#diff-25e5b255dece6e56780e97ba758b39c29df26c8aa7f77039eddc2f88b428bdd9R90
which made WorkerProtocol.run an async def. That change caused the
AMP dispatch layer to route the responder through
Deferred.fromCoroutine -> _cancellableInlineCallbacks -> _inlineCallbacks,
which sets sniffio's current_async_library_cvar to "twisted" in a
copy_context(). Since the entire test suite execution (suite.run) happens
inside context.run(gen.send, result), all tests -- including
SynchronousTestCase tests -- inherit that context variable, causing
sniffio tests that assert AsyncLibraryNotFoundError to fail under
trial -j (parallel) mode.

The fix converts WorkerProtocol.run back to a regular def that returns
a Deferred. The test execution (suite.run) is entirely synchronous and
now runs in the plain call stack with no coroutine context. The async
post-test reporting (DeferredList + sequential fallback error reporting)
is handled via Deferred callback chains, preserving the original
sequential error-handling behavior.

altendky committed:
Simplify _processRunResults to async def

Keep the error-handling loop as async def (matching trunk) since it
only runs after test execution is complete. The sync def on run()
is what matters -- it ensures suite.run() happens outside any
coroutine context. The async _processRunResults is called via
Deferred.fromCoroutine in a callback, which is fine because tests
are already done by that point.

altendky committed:
tidy

reubano commented:
Putting related issues/PRs here for reference/visibility

https://github.com/twisted/twisted/issues/9815
https://github.com/twisted/twisted/pull/1538
https://github.com/python-trio/sniffio/issues/51
https://github.com/python-trio/sniffio/pull/10
https://github.com/python-trio/sniffio/pull/13
